File: QT/cookinglistbuilder.cpp
#include "cookinglistbuilder.h"

CookingListBuilder::CookingListBuilder() {
    cooking_list = QList<Recipe>();
    ingredients_list = QList<Ingredient>();
}

void CookingListBuilder::addRecipe(Recipe recipe) {
    cooking_list.append(recipe);
}

void CookingListBuilder::removeRecipe(int index) {
    cooking_list.removeAt(index);
}

void CookingListBuilder::removeRecipe(const QString& recipe_name) {
    int size = cooking_list.size();
    for (int i=0; i<size; i++) {
        if (cooking_list[i].name == recipe_name) {
            cooking_list.removeAt(i);\
            break;
        }
    }
}

QList<Ingredient> CookingListBuilder::getIgredientsList() {
    createSummedUpIngredientsList();
    sortIngredientList();
    auto  result = ingredients_list;
    ingredients_list.clear();
    return result;
}

void CookingListBuilder::createSummedUpIngredientsList() {
    for (const auto& recipe : cooking_list) {
        sumUpIngredients(recipe.ingredients);
    }
}

void CookingListBuilder::sumUpIngredients(const QList<Ingredient>& ingredients) {
    for (const auto& current_ingredient : ingredients) {
        addIngredientToSummedList(current_ingredient);
    }
}

void CookingListBuilder::addIngredientToSummedList(const Ingredient& current_ingredient) {
    for (auto& summed_ingredient : ingredients_list) {
        if (summed_ingredient == current_ingredient) {
            summed_ingredient.amount += current_ingredient.amount;
            return;
        }
    }
    ingredients_list.append(current_ingredient);
}

void CookingListBuilder::sortIngredientList() {
    std::sort(ingredients_list.begin(),
              ingredients_list.end(),
              [&](const Ingredient& ingredient1, const Ingredient& ingredient2)
    { return ingredient1.name < ingredient2.name; }
    );
}

QStringList CookingListBuilder::getIngredientsAsStringList() {
    createSummedUpIngredientsList();
    sortIngredientList();
    auto result = getStringListIngredients();
    ingredients_list.clear();
    return result;
}

QStringList CookingListBuilder::getStringListIngredients() {
    QStringList result;
    for (const auto& ingredient : ingredients_list) {
        QString ingredient_string = buildIngredientString(ingredient);
        result.append(ingredient_string);
    }
    return result;
}

QString CookingListBuilder::buildIngredientString(const Ingredient& ingredient) {
    return ingredient.name + ": " + QString::number(ingredient.amount) + " " + ingredient.unit;
}

void CookingListBuilder::reset() {
    ingredients_list.clear();
    cooking_list.clear();
}

File: ASPNET/recipe-planer/recipe-planer/Models/CookingListBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace recipe_planer.Models
{
    public class CookingListBuilder
    {
        public List<Recipe> AvailableRecipes { get; set; }
        public List<Recipe> CookingList { get; set; }
        public List<Ingredient> SummedIngredients { get; set; }

        public CookingListBuilder()
        {
            CookingList = new List<Recipe>();
            SummedIngredients = new List<Ingredient>();
        }

        public void addRecipe(Recipe recipe)
        {
            CookingList.Add(recipe);
        }
       
        public List<Ingredient> sumUpIngredients()
        {
            SummedIngredients.Clear();
            createSummedUpIngredientsList();
            sortIngredientList();
            return SummedIngredients;
        }

        private void createSummedUpIngredientsList()
        {
            foreach(var recipe in CookingList)
            {
                sumUpIngredients(recipe.Ingredients);
            }
        }

        private void sumUpIngredients(List<Ingredient> ingredients)
        {
            foreach (var current_ingredient in ingredients)
            {
                addIngredientToSummedList(current_ingredient);
            }
        }

        private void addIngredientToSummedList(Ingredient current_ingredient)
        {
            for (int i=0; i<SummedIngredients.Count; i++)
            {
                if (SummedIngredients[i] == current_ingredient)
                {
                    SummedIngredients[i].amount += current_ingredient.amount;
                    return;
                }
            }
            SummedIngredients.Add(current_ingredient);
        }

        private void sortIngredientList()
        {
            List<Ingredient> temp = SummedIngredients.OrderBy(o => o.name).ToList();
            SummedIngredients = temp;
        }
    }
}
